opm/simulators/WellSwitchingLogger.cpp

```diff
@@ -0,0 +1,206 @@
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#if HAVE_CONFIG_H
+#include "config.h"
+#endif // HAVE_CONFIG_H
+
+#include <opm/simulators/WellSwitchingLogger.hpp>
+
+namespace Opm
+{
+namespace wellhelpers
+{
+
+#if HAVE_MPI
+int WellSwitchingLogger::calculateMessageSize(std::vector<int>& well_name_length)
+{
+
+    // Each process will send a message to the root process with
+    // the following data:
+    // total number of switches, for each switch the length of the
+    // well name, for each switch the well name and the two controls.
+    well_name_length.reserve(switchMap_.size());
+
+    for(const auto& switchEntry : switchMap_)
+    {
+        int length = switchEntry.first.size() +1;  //we write an additional \0
+        well_name_length.push_back(length);
+    }
+
+    // compute the message size
+    int message_size = 0;
+    int increment = 0;
+    // number of switches
+    MPI_Pack_size(1, MPI_INT, MPI_COMM_WORLD, &message_size);
+    message_size += increment;
+    // const char* length include delimiter for each switch
+    MPI_Pack_size(switchMap_.size(), MPI_INT, MPI_COMM_WORLD, &increment);
+    message_size += increment;
+    // for each well the name + two controls in one write
+    auto length = well_name_length.begin();
+
+    for(const auto& switchEntry : switchMap_)
+    {
+        // well name
+        MPI_Pack_size(*length, MPI_CHAR, MPI_COMM_WORLD, &increment);
+        message_size += increment;
+        // controls
+        MPI_Pack_size(2, MPI_CHAR, MPI_COMM_WORLD, &increment);
+        message_size += increment;
+        ++length;
+    }
+    return message_size;
+}
+
+void WellSwitchingLogger::packData(std::vector<int>& well_name_length,
+                                   std::vector<char>& buffer)
+{
+    // Pack the data
+    // number of switches
+    int offset = 0;
+    int no_switches = switchMap_.size();
+    MPI_Pack(&no_switches, 1, MPI_INT, buffer.data(), buffer.size(),
+             &offset, MPI_COMM_WORLD);
+    MPI_Pack(well_name_length.data(), well_name_length.size(),
+             MPI_INT, buffer.data(), buffer.size(),
+             &offset, MPI_COMM_WORLD);
+
+    for(const auto& switchEntry : switchMap_)
+    {
+        // well name
+        auto& well_name = switchEntry.first;
+        MPI_Pack(const_cast<char*>(well_name.c_str()), well_name.size()+1,
+                 MPI_CHAR, buffer.data(), buffer.size(),
+                 &offset, MPI_COMM_WORLD);
+
+        // controls
+        MPI_Pack(const_cast<char*>(switchEntry.second.data()), 2 , MPI_CHAR,
+                 buffer.data(), buffer.size(), &offset, MPI_COMM_WORLD);
+    }
+}
+
+void WellSwitchingLogger::unpackDataAndLog(std::vector<char>& recv_buffer,
+                                           const std::vector<int>& displ)
+{
+    for(int p=1; p < cc_.size(); ++p)
+    {
+        int offset = displ[p];
+        int no_switches = 0;
+        MPI_Unpack(recv_buffer.data(), recv_buffer.size(), &offset,
```

Comment:
There is no error checking here, it would be good to verify that the MPI ops were successful before accessing the results. However, throwing should not be done as that could go down the rabbit hole (being called from a destructor).

Also: I assume offset get automatically bumped by each call to MPI_Unpack()?